### PR TITLE
Attribute to mark immediate types

### PR DESCRIPTION
**Pull request by willcrichton**
This pull request adds support for a new attribute `[@@immediate]` which marks a type as a immediate, or a non-pointer. 

Types marked as immediates won't activate the GC's write barrier upon mutation (as they are not pointers and cannot induce a new pointer from the major to minor heap), which can significantly boost performance in programs relying heavily on mutable state. This feature already exists in the compiler, but it currently does not extend to abstract types. For example:

``` ocaml
module Foo : sig type t val x : t ref end = struct
  type t = int
  let x = ref 0
end
```

In the above module, modifications to `Foo.x` will invoke the write barrier because the compiler doesn't know that `t` is necessarily immediate, even though it is implemented as such. However, if you update the module to reflect that:

``` ocaml
module Foo : sig type t [@@immediate] val x : t ref end = struct
  type t = int
  let x = ref 0
end
```

Then subsequent mutations to `Foo.x` will not call the write barrier. For example, running such mutations in a tight loop:

``` ocaml
for i = 0 to 100_000_000 do
  Foo.x := !Foo.x
done
```

On my machine, the version without the attribute takes 0.51s and the version with the attribute takes 0.14s.

The feature was designed similarly to how variance is implemented in the typechecker, albeit substantially simpler. One can add a check on a type alias or definition:

``` ocaml
type t = string [@@immediate]
```

Which will produce the error message:

```
File "testsuite/tests/typing-immediate/immediate.ml", line 57, characters 2-31:
Error: Types marked with the immediate attribute must be
   non-pointer types like int or bool
```

And one can assert the immediacy of an abstract type, which will be checked during ascription:

``` ocaml
module Foo : sig type t [@@immediate] end = struct
  type t = string
end
```

Will cause the error:

```
File "testsuite/tests/typing-immediate/immediate.ml", line 67, characters 42-70:
Error: Signature mismatch:
   Modules do not match:
     sig type t = string end
   is not included in
     sig type t [@@immediate] end
   Type declarations do not match:
     type t = string
   is not included in
     type t [@@immediate]
   File "testsuite/tests/typing-immediate/immediate.ml", line 67, characters 15-35:
     Expected declaration
   File "testsuite/tests/typing-immediate/immediate.ml", line 68, characters 2-17:
     Actual declaration
   the first is not an immediate type.
```

### CODE REVIEW

File: testsuite/tests/typing-immediate/immediate.ml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+(* VALID DECLARATIONS *)
```
Comment (gasche) [inline testsuite/tests/typing-immediate/immediate.ml:4]: You have no tests where the module signature is separate from the module implementation, for example:

```
module type S = sig
  type t [@@immediate]
end

module M_valid : S = struct type t = int end
module M_invalid : S = struct type t = string end

module F (M : S) : S = M

module FM_valid = F (struct type t = int end)
module FM_invalid = F (struct type t = string end)
```

**Comment by andrejbauer** [comment]
The first example is broken. Is it supposed to have `t = int ref` instead of `t = int`?

**Comment by andrejbauer** [comment]
I would just like to notice that this looks a lot like a small fragment of an effect system for types.

**Comment by willcrichton** [comment]
@andrejbauer apologies, that was a typo. You also couldn't have `t = int ref` since that would make `t` a pointer (and hence not an immediate). And interesting observation on the effect system!

**Comment by andrejbauer** [comment]
If it was a typo, what is the correct example then? I don't get it, I think. Are you just propagating the information that a particular abstract type is "immediate" by hand?

**Comment by gasche** [comment]
My understanding is that "is immediate" is an intensional property of type that was, so far, not expressible in type declarations/interfaces. This `[@@immediate]` marker is a way to express immediateness in type signatures, and thus to propagate this information through abstraction/functorization barriers.

I think the only property of types (besides their definition, as the upper bound of the information lattice) that is expressible in signatures today is variance. We lack the way to express [injectivity](www.math.nagoya-u.ac.jp/~garrigue/papers/injectivity.pdf), [contractivity](http://caml.inria.fr/mantis/view.php?id=5863), the fact of being an upper- or lower-bound of the subtyping lattice, immediateness, [floatiness](http://caml.inria.fr/mantis/view.php?id=5361) (those two are only motivated by performance rather than expressivity). Dependent type theories typically also try to propagate (strict) productivity or size information.

I'm not sure we should consider this "propagating information by hand": it needs to be marked explicitly in the signature of abstract types (because only the programmer can decide on the intent of exposing or not those informations, making a compromise between freedom to change implementation vs. information provided by the interface), but then it is propagated by the type-checker in the normal way -- until the next abstraction boundary.

**Comment by willcrichton** [comment]
@andrejbauer I updated the example in the PR comment above. And @gasche explains the reasoning quite well!

File: testsuite/tests/typing-immediate/immediate.ml (hunk 1 above)
Comment (willcrichton) [inline testsuite/tests/typing-immediate/immediate.ml:4]: I've updated the file with the suggested tests. Thanks!

**Comment by lpw25** [comment]
> I'm not convinced by this proposition. For Immediateness and floatiness at module boundary, I think we should let the compiler do it automatically, ie put the information inside the .cmx

Putting type information in `.cmx` files would be a very significant change to the compiler, so it probably won't happen soon. Even if the information is available in `.cmx` the attribute is still useful to state the programmer's intent (in the same way that an `[@inline]` attribute would be useful).

So this patch gives us a useful feature in a much shorter time-frame -- and will remain useful even when the compiler's automatic handling of immediacy is improved.

**Comment by lpw25** [comment]
> New properties on types are for me only necessary if that allows to type more terms.

New properties on types are also useful when they allow us to type fewer terms, allowing more constraints on implementations to be expressed. This patch is an example of such a property.

**Comment by gasche** [comment]
The idea of having the information available at the source level seems good to me (better than having it only in a `.cmx`), as it allows programmers to explicitly reason about this performance characteristic -- it's the same reasoning for `[@tailcall]` for example.

**Comment by alainfrisch** [comment]
I have no strong opinion about the underlying idea (but see below), but I don't think it should be implemented with attributes.  Attributes are supposed to be ignored/passed-through by the type-checker (or impact warnings at most).  Here we are talking about a property which would have to be enforced unconditionally by the type-checker; i.e. it becomes a language feature.  We should refrain from using attributes for such extensions, and introduce a proper syntax.  Now, it could be the case that we want neither a new keyword, nor a weird mix of operator symbols; perhaps we should add once and for all a notion of modifiers on (type) declarations.

Another concern about the proposal is that it exposes something which could be implementation dependent.  What is exactly an immediate value?  Is a sum type with only constant constructor counted as such?  Also, there could be implementations of OCaml with unboxed floats (e.g. using nan-boxing techniques).  Until now, the language carefully avoided to expose data representation too much in the type system.  I'm not opposed on principle of doing so (e.g. to get better control of memoy layout), but this should be carefully designed.

A different direction to address the need covered by the proposal would be to add a kind of "super private" type declaration, where we only expose a concrete type definition to allow optimizations, but the type is otherwise considered abstract.  This does not allow e.g. to implement a functor taking an unknown type marked as "immediate", but is that an important use case?

**Comment by lpw25** [comment]
> I don't think it should be implemented with attributes. Attributes are supposed to be ignored/passed-through by the type-checker (or impact warnings at most). 

I think this proposal is a good use of attributes precisely because it is implementation-specific rather than a property of the language itself.

It is not the attribute that causes an error from the type-checker it is that the mismatch between the (implementation-specific) representation chosen for the type. It is unfortunate that the chosen internal representation affects the type-system -- but this is already the case:

```
module M : sig
    type t
    type s = { x : t }
  end = struct
    type t = float
    type s = { x : t }
  end;;
            Characters 57-111:
  ........struct
      type t = float
      type s = { x : t }
    end..
Error: Signature mismatch:
       Modules do not match:
         sig type t = float type s = { x : t; } end
       is not included in
         sig type t type s = { x : t; } end
       Type declarations do not match:
         type s = { x : t; }
       is not included in
         type s = { x : t; }
       Their internal representations differ:
       the first declaration uses unboxed float representation.
```

Also, in both the immediate and unboxed float case, the chosen representation affects the bytecode produced by ocamlc, so alternative back-ends like js_of_ocaml still need to obey the restriction to a certain extent (or reimplement their own type-based optimisation framework somehow).

**Comment by alainfrisch** [comment]
> It is not the attribute that causes an error from the type-checker it is that the mismatch between the (implementation-specific) representation chosen for the type.

Well, if you remove the attribute, the type-checker is happy.  It's hard to argue that it's not the attribute that causes the error.

Here is a specific scenario, which might be a little bit articial for now, but which could illustrate potential future issues with this approach:  imagine a library foo that documents some `[@@immediate]` types in its API.  Now, for some reason, people might want to link an alternative version foo_js.cma when building a bytecode program to be processed by js_of_ocaml (e.g. to use more efficient JS data types).  Of course, for signature checksum to match, they will need to compile against the same .cmi as the original library (and it would put a lot of burden on the build system if they have to recompile all other libraries they use and which depend on foo).  But in the alternative implementation, the implementation of the same types won't be recognized as `immediate` by the compiler (even though we don't care when compiling for js_of_ocaml).

(My proposal with "super private" types is even worse in this scenario, but people will probably hesitate more to use them in library APIs.)

**Comment by gasche** [comment]
I'm not personally bothered with the uses of attributes here. because I would consider that the attribute does not modify the semantics of programs (... those that are accepted when the attribute is present). It is true that this depends on which notion of observable semantics you have (whole programs or linkable compilation units?), and this attribute may affect the `.cmi`; but then any change to the compilation changes the `.cmx`, and I think we accept that for attributes (for [@trmc](https://github.com/ocaml/ocaml/pull/181) for example).

In Alain's example (a different compiler does not accept as immediate some type marked as immediate in the interface), I think that it is a good thing that there is a failure: the `.mli` provider expressed their intent to be warned if the performance characteristics of this code change. (And indeed, `[@immediate]` makes the code less portable, but that is by design). I'm not sure if "failing more" should prevent something from being an attribute.

Put in another way: `[@immediate]` is typical a compiler pragma: something partly compiler-implementation-defined that affects compilation behavior. It probably does not make sense to give it a "serious language syntax" for now (at least it does not make much sense in the imaginary formal semantics as a typed lambda-calculus), yet having it available improves the language's usability (here, robustness of performance reasoning). If attributes were not the right choice for those pragmas, what else should we use? Extensions?

**Comment by alainfrisch** [comment]
> what else should we use?

Well, in essence, you want to give the compiler ways to optimize client code using properties of the types exported as abstract.  Hence my suggestion of allowing to expose the definition without allowing code to do anything with it (except statically assert on the concrete definition); i.e. like "private" without creating a subtype relationship.  It should be easy to define in an implementation independent way how the type-checker would behave on these types, and it will avoid having to introduce more annotations.  For instance, one may want to expose the fact that a type is implemented as `float` (to benefit from the unboxed float record/array stuff) without allowing to see the underlying float.

And, could someone please explain which types would be considered as "immediate" under this proposal?

**Comment by lpw25** [comment]
> Hence my suggestion of allowing to expose the definition without allowing code to do anything with it (except statically assert on the concrete definition); i.e. like "private" without creating a subtype relationship.

As you said earlier, this doesn't work with functors and prevents using multiple implementations. It also makes the implementation type visible in the interface (as in it is actually written in the .mli file) which I think is undesirable.

> And, could someone please explain which types would be considered as "immediate" under this proposal?

It uses the same check as is currently used in `typeopt.ml`. So it is variants with only constant constructors, as well as `int` and `char`.

**Comment by alainfrisch** [comment]
> As you said earlier, this doesn't work with functors and prevents using multiple implementations.

Does it matter?  When you start putting annotations about the low-level representation in the interfaces, esp. something so restrictive as "immediate", you restrict the opportunities of switching implementations anyway.  Do you have concrete situations where exposing "immediate" is important and where allowing multiple implementations is relevant?

_If_ we go into the direction of specifying representation properties in the interfaces, we should think about which properties are meaningful.  For instance, being able to expose the fact that a type is not "float" would be extremely useful as well, because it allows the client code to use arrays on that type without suffering from runtime checks on floatiness.

**Comment by Chris00** [comment]
> If we go into the direction of specifying representation properties in the interfaces, [...]

I think it would also be useful for marshaling: it would be nice to have a static guarantee that _no_ exception will be raised because of a functional value hidden deep down the data structure (this is a pain — and error prone — to track manually)...

**Comment by gasche** [comment]
The patch is useful to its author, goes (I think) in the right direction from a language design point of view (make modular the properties that our users already rely on, but in a unobtrusive way because it is partly implementation-specific), is simple, short and seems to add no maintenance burden to the compiler codebase. I would be in favor of merging it.

I don't think "but we also want _this other thing_" would be a valid reason to shun the patch. If someone wants a "not a float" attribute, I would also be in favor of merging a pull request implementing it.

(Also I think Alain's suggestion to reveal the internal representation without making it usable from outside the module does not really fit the bill. There is a large quantitative difference between revealing "this is not a float" and revealing the definition. On a practical level, for soundness one would be forced to recompile the dependent modules whenever the revealed-but-not-quite definition change, whereas the "not a float" or "immediate" status is much more robust.)

**Comment by alainfrisch** [comment]
> I don't think "but we also want this other thing" would be a valid reason to shun the patch. If someone wants a "not a float" attribute, I would also be in favor of merging a pull request implementing it.

The point is that many properties might have uses, and we would need to come up with a specific patch for each of them.  Exposing the internal representation (without allowing to use it) has also the advantage of being implementation independent; it can be expressed very simply in the high-level type system and it is easy to explain to users who don't necessarily know the details of the representation.

> one would be forced to recompile the dependent modules whenever the revealed-but-not-quite definition change

I don't see that as a problem, in general.  In this specific case, people who are going to use this attribute will be sensitive to performances: they will most likely use ocamlopt, which already forces to recompile dependencies when the implementation change.

**Comment by lpw25** [comment]
> > As you said earlier, this doesn't work with functors and prevents using multiple implementations.
> 
> Does it matter?

I don't have a specific example to hand, but supporting multiple implementations which change from:

``` ocaml
type t = Foo | Bar
```

to:

``` ocaml
type t = Foo | Bar | Baz
```

seems pretty useful.

**Comment by c-cube** [comment]
Is it theoretically possible to annotate other types with such attributes? Polymorphic arrays of immediate values (or non-float values) would be nice to have, for instance.

**Comment by lpw25** [comment]
> Is it theoretically possible to annotate other types with such attributes? Polymorphic arrays of immediate values (or non-float values) would be nice to have, for instance.

This would require a kind system, which is something OCaml deliberately avoids.

**Comment by gasche** [comment]
We discussed it at the developer meeting today. I supported the design and general approach, and there seemed to be consensus that we _could_ merge it in trunk if there was solid evidence that this improves program performance.

@xavierleroy thus asked for a convincing example of use of this feature in a representative OCaml program, with information on the performance effect of the change. @willcrichton  or anyone else, would you be able to provide such a representative example (not a micro bench-mark)? The feature freeze for the next release should be in about a month, so you would need to hurry if you want it in the next release.

**Comment by yminsky** [comment]
@xavierleroy Xavier, do you need to see the actual source, or are performance numbers enough?  We can probably generate performance results pretty easily, but much of it is for source that we can't easily release.

**Comment by xavierleroy** [comment]
Just some performance results, that would be perfect.  Thanks!

**Comment by damiendoligez** [comment]
@yminsky do you have numbers now?

**Comment by yminsky** [comment]
Apologies.  We let this slip, but @lpw25 will throw together a benchmark very shortly.

**Comment by lpw25** [comment]
We have an in memory database thing that uses integers for keys and some other stuff. This is used in production (i.e. not a toy example). Benchmarks for its main operations, without `= private int`:

```
┌─────────────────────────┬──────────┬─────────┬────────────┬───────────┬────────────┐
│ Name                    │ Time/Run │ mWd/Run │   mjWd/Run │  Prom/Run │ Percentage │
├─────────────────────────┼──────────┼─────────┼────────────┼───────────┼────────────┤
│ A                       │  22.58ms │  2.21kw │ 1_804.52kw │   942.19w │     46.66% │
│ B                       │  31.29ms │  2.21kw │ 1_804.52kw │   942.27w │     64.66% │
│ C                       │  25.39ms │  2.93kw │ 1_804.96kw │ 1_383.22w │     52.45% │
│ D                       │  38.79ms │  2.93kw │ 1_804.96kw │ 1_383.31w │     80.15% │
│ E                       │  33.26ms │  3.61kw │ 1_805.52kw │ 1_938.27w │     68.73% │
│ F                       │  48.40ms │  3.61kw │ 1_805.52kw │ 1_938.41w │    100.00% │
│ G                       │  12.68ms │  2.98kw │   131.09kw │    21.10w │     26.20% │
└─────────────────────────┴──────────┴─────────┴────────────┴───────────┴────────────┘
```

and with `= private int`:

```
┌─────────────────────────┬──────────┬─────────┬────────────┬───────────┬────────────┐
│ Name                    │ Time/Run │ mWd/Run │   mjWd/Run │  Prom/Run │ Percentage │
├─────────────────────────┼──────────┼─────────┼────────────┼───────────┼────────────┤
│ A                       │  21.11ms │  2.21kw │ 1_804.52kw │   942.18w │     51.92% │
│ B                       │  29.08ms │  2.21kw │ 1_804.52kw │   942.25w │     71.51% │
│ C                       │  23.48ms │  2.93kw │ 1_804.96kw │ 1_383.20w │     57.73% │
│ D                       │  32.88ms │  2.93kw │ 1_804.96kw │ 1_383.27w │     80.86% │
│ E                       │  26.37ms │  3.61kw │ 1_805.52kw │ 1_938.22w │     64.84% │
│ F                       │  40.67ms │  3.61kw │ 1_805.52kw │ 1_938.34w │    100.00% │
│ G                       │  10.78ms │  2.98kw │   131.09kw │    21.17w │     26.50% │
└─────────────────────────┴──────────┴─────────┴────────────┴───────────┴────────────┘
```

which gives an average improvement of just under 15%. `[@@immediate]` would allow us to keep this improvement without exposing the internal int type in the interfaces.

**Comment by mshinwell** [comment]
@xavierleroy Are you satisfied with these numbers?

**Comment by damiendoligez** [comment]
Travis is failing, probably because you need to update the `.depend` files.

I talked with @xavierleroy and we're both OK to merge this.

**Comment by mshinwell** [comment]
@diml is going to do the rebase and then push it directly.

**Comment by ghost** [comment]
I squashed the commits and rebased them on the current trunk. We'll reread the result tomorrow

**Comment by ghost** [comment]
Merged manually (50dd38d4b6a992403972f37ef0ea0070dae78662)

**Pull request closed without merging**